File: rdkit-sys/wrapper/src/fingerprint.cc
#include "rust/cxx.h"
#include <DataStructs/ExplicitBitVect.h>
#include <DataStructs/BitOps.h>
#include <GraphMol/Fingerprints/Fingerprints.h>
#include <GraphMol/Fingerprints/MorganFingerprints.h>


namespace RDKit {
    std::shared_ptr<ExplicitBitVect> get_rdkit_fingerprint(const std::shared_ptr<ROMol> &mol) {
        return std::shared_ptr<ExplicitBitVect>(RDKFingerprintMol(*mol));
    }

    std::shared_ptr<ExplicitBitVect> get_morgan_fingerprint(const std::shared_ptr<ROMol> &mol, long unsigned int radius, long unsigned int n_bits) {
        return std::shared_ptr<ExplicitBitVect>(RDKit::MorganFingerprints::getFingerprintAsBitVect(*mol, radius, n_bits));
    }

    std::unique_ptr<std::vector<unsigned char>> explicit_bit_vect_to_binary_vect(const std::shared_ptr<ExplicitBitVect> &bitvect) {
        std::string binary_text = BitVectToBinaryText(*bitvect);

        std::vector<unsigned char> *vect = new std::vector<unsigned char>;
        vect->insert(vect->begin(), binary_text.begin(), binary_text.end());

        return std::unique_ptr<std::vector<unsigned char>>(vect);
    }

    std::shared_ptr<ExplicitBitVect> copy_explicit_bit_vect(const std::shared_ptr<ExplicitBitVect> &orig) {
        std::shared_ptr<ExplicitBitVect> fingerprint(new ExplicitBitVect(*orig));
        return fingerprint;
    }

    std::unique_ptr<std::vector<uint64_t>> explicit_bit_vect_to_u64_vec(const std::shared_ptr<ExplicitBitVect> &bitvect) {
        std::vector<uint64_t> bytes;
        bytes.reserve(bitvect->dp_bits->num_blocks());
        boost::to_block_range(*bitvect->dp_bits, (std::back_inserter(bytes)));
        std::vector<uint64_t> *bytes_heap = new std::vector<uint64_t>(bytes);
        return std::unique_ptr<std::vector<uint64_t>>(bytes_heap);
    }
}

File: rdkit-sys/src/bridge/fingerprint.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {

    unsafe extern "C++" {
        include!("wrapper/include/ro_mol.h");
        include!("wrapper/include/fingerprint.h");

        pub type ROMol = crate::ro_mol_ffi::ROMol;
        pub type ExplicitBitVect;
        pub fn get_rdkit_fingerprint(mol: &SharedPtr<ROMol>) -> SharedPtr<ExplicitBitVect>;
        pub fn get_morgan_fingerprint(
            mol: &SharedPtr<ROMol>,
            radius: usize,
            n_bits: usize,
        ) -> SharedPtr<ExplicitBitVect>;

        pub fn explicit_bit_vect_to_binary_vect(
            fingerprint: &SharedPtr<ExplicitBitVect>,
        ) -> UniquePtr<CxxVector<u8>>;

        pub fn copy_explicit_bit_vect(
            fingerprint: &SharedPtr<ExplicitBitVect>,
        ) -> SharedPtr<ExplicitBitVect>;

        pub fn explicit_bit_vect_to_u64_vec(
            bitvect: &SharedPtr<ExplicitBitVect>,
        ) -> UniquePtr<CxxVector<u64>>;
    }
}
